File: 415-add_strings.py
"""
https://leetcode.com/problems/add-strings/

Strat:
    Essentially writing `return str(int(num1) + int(num2))`. Use 2 pointers that 
    start from the right side of num1 and num2. First, you'll add the ones, then
    tens, then hundreds, etc. Note: the carry logic can be tricky.
    
    
Stats: O(n) / linear time, O(n) / linear space -- n is max(len(num1), len(num2))
    Runtime: 64 ms, faster than 16.05% of Python online submissions for Add Strings.
    Memory Usage: 13.4 MB, less than 99.54% of Python online submissions for Add Strings.
"""
class Solution(object):
    def addStrings(self, num1, num2):
        """
        :type num1: str
        :type num2: str
        :rtype: str
        """
        NUMS = {"1":1,"2":2,"3":3,"4":4,"5":5,"6":6,"7":7,"8":8,"9":9,"0":0}

        p1 = len(num1) - 1
        p2 = len(num2) - 1
        
        result = []
        carry = False
        
        while p1 >= 0 or p2 >= 0 or carry:
            curr_sum = 0
            
            if p1 >= 0:
                curr_sum += NUMS[num1[p1]]
                p1 -= 1
                
            if p2 >= 0:
                curr_sum += NUMS[num2[p2]]
                p2 -= 1
            
            curr_sum += carry
            
            if curr_sum >= 10: 
                #carry again
                carry = True
                curr_sum -= 10
            else:
                #no carry
                carry = False
                
            result.append(str(curr_sum))
            
        #reverse result
        result = result[::-1]
        
        #join chars into string
        return "".join(result)

File: 415-add_strings.cpp
class Solution {
public:
//     string addStrings(string num1, string num2) {        
//         int p1 = num1.size();
//         int p2 = num2.size();
              
//         string result = "";
//         bool carry = false;
        
//         while (p1 >= 0 || p2 >=0 || carry) {
//             int sum = 0;
            
//             if (p1 >= 0) {
//                 sum = sum + num1[p1] - '0';
//                 p1--;
//             }
            
//             if (p2 >= 0) {
//                 sum = sum + num1[p2] - '0';
//                 p2--;
//             }
            
//             if (sum + carry >= 10) {
//                 carry = true;
//                 sum = sum - 10;
//             }
//             else {
//                 carry = false;
//             }
            
//             result += to_string(sum);            
//         }
        
//         reverse(result.begin(), result.end());
//         return result;
//     }
    
    // not sure why my version above doesn't work; this solution is from
    // https://leetcode.com/problems/add-strings/discuss/501850/Easy-and-clean-code
    string addStrings(string num1, string num2) {        
        int idx1 = num1.size() - 1, idx2 = num2.size() - 1;
        int carry(0);
        string ans;

        while(idx1 > -1 || idx2 > -1 || carry) {
            int d1 = (idx1 >= 0) ? num1[idx1--] - '0' : 0;
            int d2 = (idx2 >= 0) ? num2[idx2--] - '0' : 0;
            int sum = d1 + d2 + carry;
            carry = sum/10;
            ans += '0' + (sum % 10);    // ans = sum%10 + '0' + ans; and no reverse
        }

        reverse(ans.begin(), ans.end());
        return ans;
    }
    
};
